core/node/state_keeper/src/v26_utils.rs

```diff
@@ -0,0 +1,443 @@
+
+    // We will also double check that the assetId corresponds to the assetId of a token that is bridged from l1
+    let expected_asset_id = encode_ntv_asset_id(l1_chain_id, origin_token_address);
+
+    // The used asset id is wrong, we should not rely on the token address to be processed by the native token vault
+    if expected_asset_id != used_asset_id {
+        return None;
+    }
+
+    Some(origin_token_address)
+}
+
+async fn calculate_expected_token_address(
+    storage: &mut Connection<'_, Core>,
+    l2_legacy_shared_bridge_address: Address,
+    l1_token_address: Address,
+) -> anyhow::Result<Address> {
+    // The source of truth for this logic is `L2SharedBridgeLegacy._calculateCreate2TokenAddress`
+
+    let beacon_proxy_bytecode_hash_key = StorageKey::new(
+        AccountTreeId::new(l2_legacy_shared_bridge_address),
+        L2_LEGACY_SHARED_BRIDGE_BEACON_PROXY_BYTECODE_KEY,
+    );
+    let beacon_proxy_bytecode_hash = storage
+        .storage_web3_dal()
+        .get_value(&beacon_proxy_bytecode_hash_key)
+        .await?;
+
+    let beacon_address_key = StorageKey::new(
+        AccountTreeId::new(l2_legacy_shared_bridge_address),
+        L2_LEGACY_SHARED_BRIDGE_UPGRADEABLE_BEACON_ADDRESS_KEY,
+    );
+    let beacon_address = storage
+        .storage_web3_dal()
+        .get_value(&beacon_address_key)
+        .await?;
+    let beacon_address = h256_to_address(&beacon_address);
+
+    let params = Create2DeploymentParams {
+        salt: address_to_h256(&l1_token_address),
+        bytecode_hash: beacon_proxy_bytecode_hash,
+        raw_constructor_input: ethabi::encode(&[
+            Token::Address(beacon_address),
+            Token::Bytes(vec![]),
+        ]),
+    };
+
+    Ok(params.derive_address(l2_legacy_shared_bridge_address))
+}
+
+/// Checks whether the token is legacy. It is legacy if both of the
+/// following is true:
+/// - It is present in legacy shared bridge
+/// - It is not present in the L2 native token vault
+async fn is_l2_token_legacy(
+    storage: &mut Connection<'_, Core>,
+    l2_legacy_shared_bridge_address: Address,
+    l2_token_address: Address,
+    expected_l1_address: Address,
+) -> anyhow::Result<bool> {
+    // 1. Read l1 token address from L2 shared bridge (must not be 0)
+
+    let stored_l1_address_key = StorageKey::new(
+        AccountTreeId::new(l2_legacy_shared_bridge_address),
+        get_address_mapping_key(&l2_token_address, L2_LEGACY_SHARED_BRIDGE_L1_ADDRESSES_KEY),
+    );
+    let stored_l1_address = storage
+        .storage_web3_dal()
+        .get_value(&stored_l1_address_key)
+        .await?;
+    let stored_l1_address = h256_to_address(&stored_l1_address);
+
+    // No address is stored, it means that the token has never been bridged before
+    // and thus, it is not legacy
+    if stored_l1_address == Address::zero() {
+        return Ok(false);
+    }
+
+    // Just for cross check
+    assert_eq!(expected_l1_address, stored_l1_address);
```

Comment:
If it fails then state keeper will panic which we don't want IMO. Should it be error log + return true as it's unexpected and we don't want to process such deposit anyway until it's investigated? Or maybe `is_l2_token_legacy` can return enum "legacy"/"not-legacy"/"unexpected-l1-address" but anyway we should treat "legacy" and "unexpected-l1-address" the same way later

Reply:
We should never hit this error in practice as it would mean a new potentially bad bug, but considering that a panic would mean downtime for everyone (and not just L1 txs) it may be better to return an error there, okay will do